src/H5T.c

```diff
@@ -1949,18 +1946,103 @@ H5Tcopy(hid_t obj_id)
         if (new_dt && H5T_close_real(new_dt) < 0)
             HDONE_ERROR(H5E_DATATYPE, H5E_CANTRELEASE, H5I_INVALID_HID, "unable to release datatype info")
 
+    FUNC_LEAVE_NOAPI(ret_value)
+} /* end H5T__copy_api_common() */
+
+/*-------------------------------------------------------------------------
+ * Function:    H5Tcopy
+ *
+ * Purpose:     Copies a datatype.  The resulting datatype is not locked.
+ *              The datatype should be closed when no longer needed by
+ *              calling H5Tclose().
+ *
+ * Return:      Success:    The ID of a new datatype
+ *
+ *              Failure:    H5I_INVALID_HID
+ *
+ * Note:        The returned type is always transient and unlocked. If the
+ *              TYPE_ID argument is a dataset instead of a datatype then
+ *              this function returns a transient, modifiable datatype
+ *              which is a copy of the dataset's datatype.
+ *
+ *-------------------------------------------------------------------------
+ */
+hid_t
+H5Tcopy(hid_t obj_id)
+{
+    H5T_t *dt        = NULL;            /* Pointer to the datatype to copy */
+    H5T_t *new_dt    = NULL;            /* Pointer to the new datatype */
+    hid_t  dset_tid  = H5I_INVALID_HID; /* Datatype ID from dataset */
+    hid_t  ret_value = H5I_INVALID_HID; /* Return value */
+
+    FUNC_ENTER_API(H5I_INVALID_HID)
+    H5TRACE1("i", "i", obj_id);
+
+    /* Copy the object's datatype synchronously */
+    if ((ret_value = H5T__copy_api_common(obj_id, NULL, NULL)) < 0)
+        HGOTO_ERROR(H5E_DATATYPE, H5E_CANTGET, H5I_INVALID_HID, "unable to copy datatype synchronously")
+
+done:
     FUNC_LEAVE_API(ret_value)
 } /* end H5Tcopy() */
 
+/*-------------------------------------------------------------------------
+ * Function:    H5Tcopy_async
```

Comment:
I'm still not entirely convinced we need an async version of H5Tcopy, since it's not _really_ an operation that deals with the file.  I see that it _could_ trigger a VOL callback, but I would defer delivering this to users until later, since I think that adding an asynchronous version of an operation that isn't really about VOL object could be a slippery slope into "asynchronous operations for *all* API routines", which is a Pandora's Box that I'd rather open later. :-)   So, I would suggest reverting the changes to H5Tcopy and just let it execute synchronously for situations where it triggered the VOL callback.  (We can always add it later)

Reply:
OK, put it simply, take everything new related to H5Tcopy out and put original H5Tcopy back, right?

Reply:
Yes :-)